File: cses/08_string_algorithms/1732_finding-borders/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    if not data:
        sys.stdout.write("\n")
        return
    s = data[0]                    # dạng bytes; s[i] là số nguyên trong Python 3
    n = len(s)

    # Prefix function (hàm thất bại của KMP), O(n) amortized:
    # pi[i] là độ dài prefix thực sự dài nhất của s[0..i] cũng là suffix của nó.
    pi = [0] * n
    k = 0
    for i in range(1, n):
        c = s[i]
        # Lùi theo chuỗi pi cho tới khi ký tự khớp hoặc về 0.
        while k > 0 and s[k] != c:
            k = pi[k - 1]
        if s[k] == c:
            k += 1
        pi[i] = k

    # Lần theo chuỗi pi từ cuối xâu để thu mọi độ dài border (giảm dần).
    res = []
    b = pi[n - 1]
    while b > 0:
        res.append(b)
        b = pi[b - 1]
    # Đảo ngược để in theo thứ tự tăng dần.
    res.reverse()

    sys.stdout.write(" ".join(map(str, res)) + "\n")


main()

File: cses/08_string_algorithms/1732_finding-borders/solution.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    std::string text;
    std::cin >> text;
    const int length = static_cast<int>(text.size());

    // Tính prefix function (hàm thất bại của KMP): prefix[i] là độ dài prefix
    // thực sự dài nhất của text[0..i] mà cũng là suffix của đoạn đó.
    std::vector<int> prefix(length, 0);
    for (int index = 1; index < length; ++index) {
        int matched = prefix[index - 1];
        // Lùi theo chuỗi pi cho tới khi ký tự khớp hoặc về 0.
        while (matched > 0 && text[index] != text[matched]) {
            matched = prefix[matched - 1];
        }
        if (text[index] == text[matched]) {
            ++matched;
        }
        prefix[index] = matched;
    }

    // Lần theo chuỗi pi từ cuối xâu để liệt kê mọi độ dài border (giảm dần).
    std::vector<int> borders;
    for (int border = prefix[length - 1]; border > 0;
         border = prefix[border - 1]) {
        borders.push_back(border);
    }
    // Đảo ngược để in theo thứ tự tăng dần theo yêu cầu.
    std::reverse(borders.begin(), borders.end());

    for (std::size_t index = 0; index < borders.size(); ++index) {
        if (index > 0) {
            std::cout << ' ';
        }
        std::cout << borders[index];
    }
    std::cout << '\n';
    return 0;
}
